src/stream/quic_lb/ngx_stream_quic_lb_retry_service.c

```diff
@@ -154,313 +154,529 @@ ngx_stream_quic_lb_gen_and_send_retry_packet(ngx_connection_t *c,
+                                      &token_body_enc, &aad, c->pool->log);
+    if (res != NGX_OK) {
+        ngx_log_error(NGX_LOG_ERR, c->pool->log, 0,
+                      "QUIC-LB, token validate faild, decrypt faild");
         return NGX_ERROR;
     }
 
-    p = pkt->token.data + iv_len;
-    len = pkt->token.len - iv_len;
-
-    if (EVP_DecryptUpdate(ctx, tdec, &len, p, len) != 1) {
-        EVP_CIPHER_CTX_free(ctx);
+#ifdef NGX_QUIC_DEBUG_CRYPTO
+    ngx_quic_hexdump(c->pool->log, "QUIC-LB, quic retry token validate, body enc:",
+                     token_body_plaintext.data, token_body_plaintext.len);
+#endif
+    /* you can implement your self define token validation here */
+    res = ngx_stream_quic_lb_parse_quic_lb_model_plain_token_body(token_body_plaintext.data,
+                                                                  token_body_plaintext.len, &token_body);
+    if (res != NGX_OK) {
+        ngx_log_error(NGX_LOG_ERR, c->pool->log, 0,
+                      "QUIC-LB, token validate faild, parse token body faild");
         return NGX_ERROR;
     }
 
-    if (EVP_DecryptFinal_ex(ctx, tdec + len, &tlen) <= 0) {
-        EVP_CIPHER_CTX_free(ctx);
+    res = ngx_stream_quic_lb_validate_timestamp(token_body.expire_time);
+    if (res != NGX_OK) {
+        ngx_log_error(NGX_LOG_ERR, c->pool->log, 0,
+                      "QUIC-LB, token validate faild, token expired");
         return NGX_ERROR;
     }
 
-    EVP_CIPHER_CTX_free(ctx);
+    return NGX_OK;
+}
 
-    switch (c->sockaddr->sa_family) {
 
-#if (NGX_HAVE_INET6)
-    case AF_INET6:
-        sin6 = (struct sockaddr_in6 *) c->sockaddr;
-        len = sizeof(struct in6_addr);
-        data = sin6->sin6_addr.s6_addr;
-        break;
-#endif
+static uint64_t
+ngx_stream_quic_lb_get_timestamp(uint64_t token_alive_time)
+{
+    time_t      now;
+    uint64_t    expire_time;
 
-#if (NGX_HAVE_UNIX_DOMAIN)
-    case AF_UNIX:
+    now = time(NULL);
+    if (now < 0) {
+        return NGX_ERROR;
+    }
 
-        len = ngx_min(c->addr_text.len, NGX_QUIC_MAX_TOKEN_SIZE - sizeof(msec));
-        data = c->addr_text.data;
-        break;
-#endif
+    expire_time = now;
+    expire_time += token_alive_time;
 
-    default: /* AF_INET */
-        sin = (struct sockaddr_in *) c->sockaddr;
+    return expire_time;
+}
 
-        len = sizeof(in_addr_t);
-        data = (u_char *) &sin->sin_addr;
-        ngx_log_error(NGX_LOG_DEBUG, c->pool->log, 0,
-                      "QUIC-LB, retry service, addr value in token is:%d", sin->sin_addr);
-        break;
-    }
 
-    if (ngx_memcmp(tdec, data, len) != 0) {
+static ngx_int_t
+ngx_stream_quic_lb_validate_timestamp(uint64_t expire_time)
+{
+    time_t      now;
+
+    now = time(NULL);
+    if (now < 0) {
         return NGX_ERROR;
     }
 
-    ngx_memcpy(&msec, tdec + len, sizeof(msec));
-
-    if (ngx_current_msec - msec > NGX_QUIC_RETRY_LIFETIME) {
-        return NGX_ERROR;
+    if (now > 0 && (uint64_t)now > expire_time) {
```

Comment:
"in order to allow for clock skew, it SHOULD NOT
consider tokens to be expired if the Timestamp encodes a few seconds
in the past"

maybe need  some seconds as buffer

Reply:
draft said that `quic-server` and `quic-lb` should have a uniform time,  clock skew seems always been the problem of those cookie like mechanism, we need not to avoid it here